addons/sale_timesheet/models/account.py

```diff
@@ -55,12 +70,62 @@ def _get_timesheet_cost(self, values):
             amount = -unit_amount * cost
             amount_converted = employee.currency_id.compute(amount, analytic_account.currency_id)
 
-            return {
+            result.update({
                 'amount': amount_converted,
                 'product_uom_id': uom.id,
                 'account_id': account_id,
-            }
-        return {}
+            })
+        return result
+
+    def _get_timesheet_billing_values(self, values):
+        """
+            If invoice on delivered quantity:
+                timesheet hours * (SO Line Price) * (1- discount),
+            elif invoice on ordered quantities & create task:
+                min (
+                    timesheet hours * (SO Line unit price) * (1- discount),
+                    TOTAL SO - TOTAL INVOICED - sum(timesheet revenues with invoice_id=False)
+                )
+            else:
+                0
+        """
+        result = {}
+        unit_amount = values.get('unit_amount', 0.0) or self.unit_amount
+        so_line_id = values.get('so_line') or self.so_line.id
+        if not self.env.context.get('create'):
```

Comment:
--> if self.context.get() -> return, and lessen indentation. We will clean that key brol another day. Maybe. Not.